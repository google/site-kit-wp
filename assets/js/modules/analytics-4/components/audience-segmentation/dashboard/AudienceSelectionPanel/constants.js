/**
 * AudienceSelectionPanel components - constants.
 *
 * Site Kit by Google, Copyright 2024 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

export const AUDIENCE_ADD_GROUP_NOTICE_SLUG =
	'audience-segmentation-add-group-notice';
export const AUDIENCE_SELECTION_PANEL_OPENED_KEY =
	'googlesitekit-audience-selection-panel-opened';
export const AUDIENCE_SELECTION_FORM = 'audience-selection-form';
export const AUDIENCE_SELECTED = 'audience-selected';
export const AUDIENCE_SELECTION_CHANGED = 'audience-selection-changed';
export const AUDIENCE_CREATION_NOTICE_SLUG =
	'audience-segmentation-creation-notice';
export const AUDIENCE_CREATION_SUCCESS_NOTICE_SLUG =
	'audience-segmentation-creation-success-notice';
export const MIN_SELECTED_AUDIENCES_COUNT = 1;
export const MAX_SELECTED_AUDIENCES_COUNT = 3;
export const AUDIENCE_CREATION_EDIT_SCOPE_NOTICE_SLUG =
	'audience-creation-edit-scope-notice';
export const AUDIENCE_CREATION_FORM = 'audience-creation-form';
