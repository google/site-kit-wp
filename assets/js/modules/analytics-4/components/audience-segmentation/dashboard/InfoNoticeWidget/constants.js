/**
 * Constants related to audience segmentation info notices.
 *
 * Site Kit by Google, Copyright 2024 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

/**
 * WordPress dependencies
 */
import { __ } from '@wordpress/i18n';

/**
 * UI key to display audience segmentation info notice.
 *
 * @since 1.135.0
 * @private
 */
export const AUDIENCE_INFO_NOTICE_HIDE_UI =
	'audience-segmentation-info-notice-ui';

/**
 * Slug for audience segmentation info notice.
 *
 * @since 1.130.0
 * @private
 */
export const AUDIENCE_INFO_NOTICE_SLUG = 'audience-segmentation-info-notice';

// Notice slugs: new-visitors, compare-metrics, custom-audiences, purchasers, returning-visitors, compare-new-returning, or compare-cities

/**
 * List of info notices for audiences.
 *
 * @since 1.130.0
 * @private
 */
export const AUDIENCE_INFO_NOTICES = [
	{
		slug: 'new-visitors',
		content: __(
			'The higher the portion of new visitors you have, the more your audience is growing. Looking at what content brings them to your site may give you insights on how to reach even more people.',
			'google-site-kit'
		),
	},
	{
		slug: 'compare-metrics',
		content: __(
			'Select up to three visitor groups to display on the dashboard and easily compare metrics between them.',
			'google-site-kit'
		),
	},
	{
		slug: 'custom-audiences',
		content: __(
			'Configure your own custom audiences in Analytics to gain deeper insights into visitor behavior, for example consider creating an “Existing customers” or “Subscribers” segment, depending on what goals you have for your site.',
			'google-site-kit'
		),
	},
	{
		slug: 'purchasers',
		content: __(
			'Select the Purchasers visitor group to gain insights into which visitors bring the most revenue to your site.',
			'google-site-kit'
		),
	},
	{
		slug: 'returning-visitors',
		content: __(
			'The more returning visitors your site has, the stronger and more loyal an audience you’re building. Check which content brings people back to your site - it might help you create a strategy to build a community.',
			'google-site-kit'
		),
	},
	{
		slug: 'compare-new-returning',
		content: __(
			'Compare the ratio of “new” to “returning” visitors – this can give you insights on whether you have more people stopping by as a one-off, or more loyal visitors.',
			'google-site-kit'
		),
	},
	{
		slug: 'compare-cities',
		content: __(
			'Check the cities which bring you more new vs more returning visitors – there might be new audiences you could engage with in locations you hadn’t thought about.',
			'google-site-kit'
		),
	},
];
