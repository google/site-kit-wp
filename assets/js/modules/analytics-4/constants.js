/**
 * GA4 module constants.
 *
 * Site Kit by Google, Copyright 2022 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

export const ACTIVATION_STEP_REMINDER = 0;
export const ACTIVATION_STEP_SETUP = 1;
export const ACTIVATION_STEP_SUCCESS = 2;
export const ACTIVATION_STEP_IN_PROGRESS = 3;

export const ACTIVATION_ACKNOWLEDGEMENT_TOOLTIP_STATE_KEY =
	'activation-acknowledgement-tooltip-state';

export const GA4_ACTIVATION_BANNER_STATE_KEY = 'ga4-activation-banner-state';

export const ENHANCED_MEASUREMENT_ACTIVATION_BANNER_TOOLTIP_STATE_KEY =
	'enhanced-measurement-activation-banner-tooltip-state';
export const LEGACY_ENHANCED_MEASUREMENT_ACTIVATION_BANNER_DISMISSED_ITEM_KEY =
	'enhanced-measurement-activation-banner-dismissed-item';

export const REPORT_ARGS_SELECTED_METRIC_KEY = '_r.explorerCard..selmet';
export const REPORT_ARGS_SELECTED_DIMENSION_KEY = '_r.explorerCard..seldim';
export const REPORT_ARGS_DATA_FILTERS_KEY = '_r..dataFilters';
export const REPORT_ARGS_NAV_KEY = '_r..nav';

export const KM_CONNECT_GA4_CTA_WIDGET_DISMISSED_ITEM_KEY =
	'key-metrics-connect-ga4-cta-widget';

/**
 * Date that Site Kit will automatically switch to using GA4: September 25, 2023.
 *
 * With Site Kit's maximum date range being three months, once this period has passed since the UA cutoff date,
 * it will no longer be able to show any historical data for UA and will automatically switch to GA4.
 * For operational reasons, we automatically make the switch a few days before the three month cutoff point.
 *
 * @since 1.107.0
 * @see UA_CUTOFF_DATE.
 */
export const GA4_AUTO_SWITCH_DATE = '2023-09-25';
export const MODULE_SLUG_ANALYTICS_4 = 'analytics-4';
