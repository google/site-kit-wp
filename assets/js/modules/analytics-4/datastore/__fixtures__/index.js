/**
 * GA4 Datastore Fixtures.
 *
 * Site Kit by Google, Copyright 2021 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

export { default as accountSummaries } from './account-summaries.json';
export { default as audiences } from './audiences.json';
export { default as availableAudiences } from './available-audiences.json';
export { default as container } from './container.json';
export { default as containerE2E } from './container-e2e.json';
export { default as containerDestinations } from './container-destinations.json';
export { default as conversionEvents } from './conversion-events.json';
export { default as createProperty } from './create-property.json';
export { default as createWebDataStream } from './create-webdatastream.json';
export { default as defaultSettings } from './default-settings.json';
export { default as defaultEnhancedMeasurementSettings } from './default-enhanced-measurement-settings.json';
export { default as pageTitles } from './page-titles.json';
export { default as properties } from './properties.json';
export { default as report } from './report.json';
export { default as pivotReport } from './pivot-report.json';
export { default as webDataStreams } from './webdatastreams.json';
export { default as webDataStreamsBatch } from './webdatastreams-batch.json';
export { default as webDataStreamsBatchSetup } from './webdatastreams-batch-setup.json';
export { default as googleTagSettings } from './google-tag-settings.json';
