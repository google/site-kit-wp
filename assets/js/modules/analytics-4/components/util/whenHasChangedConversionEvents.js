/**
 * Site Kit by Google, Copyright 2025 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

/**
 * Internal dependencies
 */
import { useSelect } from 'googlesitekit-data';
import { MODULES_ANALYTICS_4 } from '../../datastore/constants';

/**
 * Higher-Order Component to render wrapped components when changed conversion events are detected.
 *
 * @since n.e.x.t
 *
 * @return {Function} Enhancing function.
 */
export default function whenHasChangedConversionEvents() {
	return ( WrappedComponent ) => {
		function WhenHasChangedConversionEvents( props ) {
			const hasChangedConversionEvents = useSelect( ( select ) => {
				const {
					hasNewConversionReportingEvents,
					hasLostConversionReportingEvents,
				} = select( MODULES_ANALYTICS_4 );

				return (
					hasNewConversionReportingEvents() ||
					hasLostConversionReportingEvents()
				);
			}, [] );

			if ( hasChangedConversionEvents ) {
				return <WrappedComponent { ...props } />;
			}

			const { WidgetNull = null } = props;
			return WidgetNull && <WidgetNull />;
		}

		return WhenHasChangedConversionEvents;
	};
}
