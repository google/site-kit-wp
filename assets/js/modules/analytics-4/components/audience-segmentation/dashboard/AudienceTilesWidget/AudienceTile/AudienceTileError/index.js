/**
 * AudienceTileError component.
 *
 * Site Kit by Google, Copyright 2024 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

/**
 * External dependencies
 */
import PropTypes from 'prop-types';

/**
 * Internal dependencies
 */
import { isInsufficientPermissionsError } from '../../../../../../../../util/errors';
import TileErrorContent from './TileErrorContent';
import withIntersectionObserver from '../../../../../../../../util/withIntersectionObserver';
import useViewContext from '../../../../../../../../hooks/useViewContext';
import { trackEvent } from '../../../../../../../../util';

const TileErrorContentWithIntersectionObserver =
	withIntersectionObserver( TileErrorContent );

export default function AudienceTileError( { audienceSlug, errors } ) {
	const viewContext = useViewContext();
	const hasInsufficientPermissionsError = errors.some( ( err ) =>
		isInsufficientPermissionsError( err )
	);

	return (
		<TileErrorContentWithIntersectionObserver
			errors={ errors }
			onInView={ () => {
				const action = hasInsufficientPermissionsError
					? 'insufficient_permissions_error'
					: 'data_loading_error';

				trackEvent(
					`${ viewContext }_audiences-tile`,
					action,
					audienceSlug
				);
			} }
			onRetry={ () => {
				trackEvent(
					`${ viewContext }_audiences-tile`,
					'data_loading_error_retry',
					audienceSlug
				);
			} }
			onRequestAccess={ () => {
				trackEvent(
					`${ viewContext }_audiences-tile`,
					'insufficient_permissions_error_request_access',
					audienceSlug
				);
			} }
		/>
	);
}

AudienceTileError.propTypes = {
	audienceSlug: PropTypes.string.isRequired,
	errors: PropTypes.array.isRequired,
};
