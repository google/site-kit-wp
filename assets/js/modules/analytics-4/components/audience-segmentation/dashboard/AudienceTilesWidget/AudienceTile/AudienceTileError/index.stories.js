/**
 * AudienceTileError Component Stories.
 *
 * Site Kit by Google, Copyright 2024 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

/**
 * Internal dependencies
 */
import { useSelect } from 'googlesitekit-data';
import AudienceTileError from '.';
import { MODULES_ANALYTICS_4 } from '../../../../../../datastore/constants';
import {
	WithTestRegistry,
	createTestRegistry,
	provideModuleRegistrations,
	provideModules,
} from '../../../../../../../../../../tests/js/utils';
import WithRegistrySetup from '../../../../../../../../../../tests/js/WithRegistrySetup';
import { ERROR_REASON_INSUFFICIENT_PERMISSIONS } from '../../../../../../../../util/errors';

function AudienceTileErrorWrapper( { ...args } ) {
	const errors = useSelect( ( select ) =>
		select( MODULES_ANALYTICS_4 ).getErrors()
	);

	return <AudienceTileError errors={ errors } { ...args } />;
}

function Template( { setupRegistry = async () => {}, ...args } ) {
	const setupRegistryCallback = async ( registry ) => {
		provideModules( registry );
		provideModuleRegistrations( registry );
		await setupRegistry( registry );
	};
	return (
		<WithRegistrySetup func={ setupRegistryCallback }>
			<AudienceTileErrorWrapper { ...args } />
		</WithRegistrySetup>
	);
}

export const Default = Template.bind( {} );
Default.storyName = 'Default';
Default.args = {
	setupRegistry: async ( registry ) => {
		await registry.dispatch( MODULES_ANALYTICS_4 ).receiveError(
			{
				code: 'test-error-code',
				message: 'Test error message',
				data: {
					reason: 'Data Error',
				},
			},
			'getReport',
			[
				{
					dimensions: [ 'ga:date' ],
					metrics: [ { expression: 'ga:users' } ],
					startDate: '2020-08-11',
					endDate: '2020-09-07',
				},
			]
		);
	},
};
Default.scenario = {
	label: 'Modules/Analytics4/Components/AudienceSegmentation/Dashboard/AudienceTileError/Default',
};

export const InsufficientPermissions = Template.bind( {} );
InsufficientPermissions.storyName = 'InsufficientPermissions';
InsufficientPermissions.args = {
	setupRegistry: async ( registry ) => {
		provideModules( registry, [
			{
				active: true,
				connected: true,
				slug: 'analytics-4',
			},
		] );

		const [ accountID, propertyID, measurementID, webDataStreamID ] = [
			'12345',
			'34567',
			'56789',
			'78901',
		];

		await registry
			.dispatch( MODULES_ANALYTICS_4 )
			.setAccountID( accountID );
		await registry
			.dispatch( MODULES_ANALYTICS_4 )
			.setPropertyID( propertyID );
		await registry
			.dispatch( MODULES_ANALYTICS_4 )
			.setMeasurementID( measurementID );
		await registry
			.dispatch( MODULES_ANALYTICS_4 )
			.setWebDataStreamID( webDataStreamID );
		await registry.dispatch( MODULES_ANALYTICS_4 ).receiveError(
			{
				code: 'test-error-code',
				message: 'Test error message',
				data: {
					reason: ERROR_REASON_INSUFFICIENT_PERMISSIONS,
				},
			},
			'getAccountID'
		);
	},
};
InsufficientPermissions.scenario = {
	label: 'Modules/Analytics4/Components/AudienceSegmentation/Dashboard/AudienceTileError/InsufficientPermissions',
};

export default {
	title: 'Modules/Analytics4/Components/AudienceSegmentation/Dashboard/AudienceTileError',
	component: AudienceTileError,
	decorators: [
		( Story ) => {
			const registry = createTestRegistry();

			return (
				<WithTestRegistry registry={ registry }>
					<Story />
				</WithTestRegistry>
			);
		},
	],
};
