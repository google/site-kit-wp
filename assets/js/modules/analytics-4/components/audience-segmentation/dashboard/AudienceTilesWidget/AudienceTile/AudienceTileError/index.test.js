/**
 * AudienceTileError component tests.
 *
 * Site Kit by Google, Copyright 2024 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

/**
 * External dependencies
 */
import { useIntersection as mockUseIntersection } from 'react-use';

/**
 * Internal dependencies
 */
import AudienceTileError from '.';
import {
	act,
	fireEvent,
	render,
} from '../../../../../../../../../../tests/js/test-utils';
import {
	createTestRegistry,
	provideModuleRegistrations,
	provideModules,
	provideSiteInfo,
	provideUserInfo,
	waitForDefaultTimeouts,
} from '../../../../../../../../../../tests/js/utils';
import { VIEW_CONTEXT_MAIN_DASHBOARD } from '../../../../../../../../googlesitekit/constants';
import { MODULES_ANALYTICS_4 } from '../../../../../../datastore/constants';
import { ERROR_REASON_INSUFFICIENT_PERMISSIONS } from '../../../../../../../../util/errors';
import * as tracking from '../../../../../../../../util/tracking';

jest.mock( 'react-use', () => ( {
	...jest.requireActual( 'react-use' ),
	useIntersection: jest.fn(),
} ) );

const mockTrackEvent = jest.spyOn( tracking, 'trackEvent' );
mockTrackEvent.mockImplementation( () => Promise.resolve() );

describe( 'AudienceTileError', () => {
	let registry;

	const audienceSlug = 'new-visitors';

	const insufficientPermissionsError = {
		code: 'test_error',
		message: 'Error message.',
		data: {
			status: 403,
			reason: ERROR_REASON_INSUFFICIENT_PERMISSIONS,
		},
	};

	const notFoundError = {
		code: 404,
		message: 'Not found or permission denied.',
		data: { status: 404, reason: 'notFound' },
	};

	const reportOptions = {
		metrics: [
			{
				name: 'totalUsers',
			},
		],
		dimensions: [
			{
				name: 'date',
			},
		],
		startDate: '2020-08-11',
		endDate: '2020-09-07',
	};

	beforeEach( () => {
		mockUseIntersection.mockImplementation( () => ( {
			isIntersecting: false,
			intersectionRatio: 0,
		} ) );

		registry = createTestRegistry();

		provideModules( registry, [
			{
				active: true,
				connected: true,
				slug: 'analytics-4',
			},
		] );
		provideModuleRegistrations( registry );
		provideSiteInfo( registry );
		provideUserInfo( registry );

		registry.dispatch( MODULES_ANALYTICS_4 ).receiveGetSettings( {
			accountID: '12345',
			propertyID: '34567',
			measurementID: '56789',
			webDataStreamID: '78901',
		} );
	} );

	afterEach( () => {
		mockTrackEvent.mockClear();
	} );

	it( 'should correctly render the insufficient permissions variant', async () => {
		await registry
			.dispatch( MODULES_ANALYTICS_4 )
			.receiveError( insufficientPermissionsError, 'getReport', [
				reportOptions,
			] );

		const errors = registry.select( MODULES_ANALYTICS_4 ).getErrors();

		const { container, getByText } = render(
			<AudienceTileError
				audienceSlug={ audienceSlug }
				errors={ errors }
			/>,
			{
				registry,
			}
		);

		expect( getByText( 'Insufficient permissions' ) ).toBeInTheDocument();

		expect( container ).toMatchSnapshot();
	} );

	it( 'should correctly render the generic error variant', async () => {
		await registry
			.dispatch( MODULES_ANALYTICS_4 )
			.receiveError( notFoundError, 'getReport', [ reportOptions ] );

		const errors = registry.select( MODULES_ANALYTICS_4 ).getErrors();

		const { container, getByText, waitForRegistry } = render(
			<AudienceTileError
				audienceSlug={ audienceSlug }
				errors={ errors }
			/>,
			{
				registry,
			}
		);

		await waitForRegistry();

		expect( getByText( 'Data loading failed' ) ).toBeInTheDocument();

		expect( container ).toMatchSnapshot();
	} );

	it( 'should track an event when the insufficient permissions error variant is viewed', async () => {
		const { rerender, waitForRegistry } = render(
			<AudienceTileError
				audienceSlug={ audienceSlug }
				errors={ [ insufficientPermissionsError ] }
			/>,
			{
				registry,
				viewContext: VIEW_CONTEXT_MAIN_DASHBOARD,
			}
		);

		expect( mockTrackEvent ).toHaveBeenCalledTimes( 0 );

		// Simulate the CTA becoming visible.
		mockUseIntersection.mockImplementation( () => ( {
			isIntersecting: true,
			intersectionRatio: 1,
		} ) );

		rerender(
			<AudienceTileError
				audienceSlug={ audienceSlug }
				errors={ [ insufficientPermissionsError ] }
			/>
		);

		await waitForRegistry();

		expect( mockTrackEvent ).toHaveBeenCalledWith(
			'mainDashboard_audiences-tile',
			'insufficient_permissions_error',
			audienceSlug
		);
	} );

	it( 'should track an event when "Request access" is clicked on the insufficient permissions error variant', async () => {
		await registry
			.dispatch( MODULES_ANALYTICS_4 )
			.receiveError( insufficientPermissionsError, 'getReport', [
				reportOptions,
			] );

		const errors = registry.select( MODULES_ANALYTICS_4 ).getErrors();

		const { getByRole } = render(
			<AudienceTileError
				audienceSlug={ audienceSlug }
				errors={ errors }
			/>,
			{
				registry,
				viewContext: VIEW_CONTEXT_MAIN_DASHBOARD,
			}
		);

		expect( mockTrackEvent ).toHaveBeenCalledTimes( 0 );

		await act( async () => {
			fireEvent.click(
				getByRole( 'button', { name: /Request access/ } )
			);

			// Allow the `trackEvent()` promise to resolve.
			await waitForDefaultTimeouts();
		} );

		expect( mockTrackEvent ).toHaveBeenCalledWith(
			'mainDashboard_audiences-tile',
			'insufficient_permissions_error_request_access',
			audienceSlug
		);
	} );

	it( 'should track an event when the generic error variant is viewed', async () => {
		const { rerender, waitForRegistry } = render(
			<AudienceTileError
				audienceSlug={ audienceSlug }
				errors={ [ notFoundError ] }
			/>,
			{
				registry,
				viewContext: VIEW_CONTEXT_MAIN_DASHBOARD,
			}
		);

		expect( mockTrackEvent ).toHaveBeenCalledTimes( 0 );

		// Simulate the CTA becoming visible.
		mockUseIntersection.mockImplementation( () => ( {
			isIntersecting: true,
			intersectionRatio: 1,
		} ) );

		rerender(
			<AudienceTileError
				audienceSlug={ audienceSlug }
				errors={ [ notFoundError ] }
			/>
		);

		await waitForRegistry();

		expect( mockTrackEvent ).toHaveBeenCalledWith(
			'mainDashboard_audiences-tile',
			'data_loading_error',
			audienceSlug
		);
	} );

	it( 'should track an event when "Retry" is clicked on the generic error variant', async () => {
		await registry
			.dispatch( MODULES_ANALYTICS_4 )
			.receiveError( notFoundError, 'getReport', [ reportOptions ] );

		const errors = registry.select( MODULES_ANALYTICS_4 ).getErrors();

		const { getByRole } = render(
			<AudienceTileError
				audienceSlug={ audienceSlug }
				errors={ errors }
			/>,
			{
				registry,
				viewContext: VIEW_CONTEXT_MAIN_DASHBOARD,
			}
		);

		expect( mockTrackEvent ).toHaveBeenCalledTimes( 0 );

		await act( async () => {
			fireEvent.click( getByRole( 'button', { name: /Retry/ } ) );

			// Allow the `trackEvent()` promise to resolve.
			await waitForDefaultTimeouts();
		} );

		expect( mockTrackEvent ).toHaveBeenCalledWith(
			'mainDashboard_audiences-tile',
			'data_loading_error_retry',
			audienceSlug
		);
	} );
} );
