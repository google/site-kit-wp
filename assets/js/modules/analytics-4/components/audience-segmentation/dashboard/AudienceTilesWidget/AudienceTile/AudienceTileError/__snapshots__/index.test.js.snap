// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AudienceTileError should correctly render the generic error variant 1`] = `
<div>
  <div
    class="googlesitekit-audience-segmentation-tile-error"
  >
    <div
      class="googlesitekit-audience-segmentation-tile-error__container"
    >
      <svg />
      <div
        class="googlesitekit-audience-segmentation-tile-error__body"
      >
        <div
          class="googlesitekit-audience-segmentation-tile-error__message"
        >
          <h3
            class="googlesitekit-audience-segmentation-tile-error__title"
          >
            Data loading failed
          </h3>
        </div>
        <div
          class="googlesitekit-audience-segmentation-tile-error__actions"
        >
          <div
            class="googlesitekit-report-error-actions"
          >
            <button
              class="mdc-button mdc-button--raised mdc-button--danger"
              target="_self"
            >
              <span
                class="mdc-button__label"
              >
                Retry
              </span>
            </button>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`AudienceTileError should correctly render the insufficient permissions variant 1`] = `
<div>
  <div
    class="googlesitekit-audience-segmentation-tile-error"
  >
    <div
      class="googlesitekit-audience-segmentation-tile-error__container"
    >
      <svg />
      <div
        class="googlesitekit-audience-segmentation-tile-error__body"
      >
        <div
          class="googlesitekit-audience-segmentation-tile-error__message"
        >
          <h3
            class="googlesitekit-audience-segmentation-tile-error__title"
          >
            Insufficient permissions
          </h3>
        </div>
        <div
          class="googlesitekit-audience-segmentation-tile-error__actions"
        >
          <div
            class="googlesitekit-report-error-actions"
          >
            <a
              aria-label="Request access (opens in a new tab)"
              class="mdc-button mdc-button--raised mdc-button--danger"
              href="https://accounts.google.com/accountchooser?continue=https%3A%2F%2Fanalytics.google.com%2Fanalytics%2Fweb%2F%23%2Fa12345p34567%2Fadmin%2Fstreams%2Ftable%2F78901&Email=wapuu.wordpress%40gmail.com"
              role="button"
              target="_blank"
            >
              <span
                class="mdc-button__label"
              >
                Request access
              </span>
            </a>
            <div>
              Contact your administrator. Trouble getting access? 
              <a
                aria-label="Get help (opens in a new tab)"
                class="googlesitekit-cta-link"
                href="https://sitekit.withgoogle.com/support/?error_id=analytics-4_insufficient_permissions"
                rel="noopener noreferrer"
                target="_blank"
              >
                <span
                  class="googlesitekit-cta-link__contents"
                >
                  Get help
                </span>
              </a>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
